app-ng/src/errors.lisp

```diff
@@ -0,0 +1,26 @@
+(in-package #:qvm-app-ng)
+
+(define-condition rpc-error (simple-error)
+  ((http-status
+    :initarg :http-status
+    :initform +http-internal-server-error+
+    :reader rpc-error-http-status))
+  (:report (lambda (condition stream)
+             (format stream "QVM RPC Error:~@[ ~A~]"
+                     (http-status-string (rpc-error-http-status condition)))
+             (alexandria:when-let ((format-control (simple-condition-format-control condition)))
+               (format stream "~&~?" format-control (simple-condition-format-arguments condition)))))
+  (:documentation "Base error type for all RPC errors."))
+
+(macrolet ((def-rpc-error (name http-status documentation)
+             `(progn
+                (define-condition ,name (rpc-error)
+                  ()
+                  (:default-initargs :http-status ,http-status)
+                  (:documentation ,documentation))
+                (defun ,name (&optional format-control &rest format-arguments)
+                  (error ',name :format-control format-control :format-arguments format-arguments)))))
+  (def-rpc-error rpc-bad-request-error +http-bad-request+
+    "Error signalled when caller makes an invalid RPC request.")
+  (def-rpc-error rpc-parameter-parse-error +http-bad-request+
```

Comment:
Would this be an HTTP error 422 error instead of a 400? That is: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422 vs. https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400.

Reply:
Hmmmm, good question. The line is sufficiently blurred that I think a case can be made for either. 422 seems more correct/precise from an HTTP standpoint, whereas 400 is maybe more generic but easier for a casual reader to understand "Bad Request" as opposed to "Unprocessable Entity".

I also like that 400 Bad Request is part of the HTTP/1 .1 spec, while 422 is apparently part of some [WebDAV extensions RFC](https://tools.ietf.org/html/rfc4918#section-11.2) (even though I know it's unlikely to cause any issue)...

I think I lean towards to sticking with Bad Request, but I can be convinced otherwise. I *do* love being pedantically correct :-). Let me chew it on.

Reply:
That makes sense.